### PR TITLE
[1847 AE] Corporations, double shares, investor shares, logos

### PR DESCRIPTION
**Pull request by Galatolol**

### CODE REVIEW

File: lib/engine/game/g_1847_ae/game.rb [hunk 1]
```diff
@@ -111,6 +123,57 @@ def operating_round(round_num)
...
+        def init_corporations(stock_market)
+          self.class::CORPORATIONS.map do |corporation|
+            par_price = stock_market.par_prices.find { |p| p.price == corporation[:required_par_price] }
+            corporation[:par_price] = par_price
+            G1847AE::Corporation.new(
+              min_price: par_price.price,
+              capitalization: self.class::CAPITALIZATION,
+              **corporation.merge(corporation_opts),
+            )
+          end
```
Comment (bentziaxl) [inline lib/engine/game/g_1847_ae/game.rb:153]: no need to do this, define CORPORATION_CLASS instead

File: lib/engine/game/g_1847_ae/game.rb [hunk 2]
```diff
@@ -111,6 +123,57 @@ def operating_round(round_num)
...
+        end
+
+        def setup
+          # Reserve investor shares and add money for them to treasury
+          [saar.shares[1], saar.shares[2], hlb.shares[1]].each { |s| s.buyable = false }
+          saar.cash += saar.par_price.price * 2
+          hlb.cash += hlb.par_price.price
+        end
```
Comment (bentziaxl) [inline lib/engine/game/g_1847_ae/game.rb:161]: do saar and hlb have a par price at the start of the game? shouldn't this code run after parring?

Comment (Galatolol) [reply lib/engine/game/g_1847_ae/game.rb:161]: They do, in this game you don't choose pars.

File: lib/engine/game/g_1847_ae/game.rb (hunk 1 above)
Comment (Galatolol) [reply lib/engine/game/g_1847_ae/game.rb:153]: I have to pass the par_price object to the Corporation. And I can't reach @game (it's nil) from inside the Corporation.initialize from what I see.

Comment (bentziaxl) [reply lib/engine/game/g_1847_ae/game.rb:153]: you can always set min_price in setup

File: lib/engine/game/g_1847_ae/share_pool.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+module Engine
+  module Game
+    module G1847AE
+      class SharePool < Engine::SharePool
+        def change_president(presidents_share, swap_to, president, _previous_president)
+          corporation = presidents_share.corporation
+          incoming_pres_shares = president.shares_of(corporation)
+
+          single_shares = incoming_pres_shares.filter { |s| s.double_cert == false }
```
Comment (tobymao) [inline lib/engine/game/g_1847_ae/share_pool.rb:11]: don't use filter, use reject

incoming_pre_share.reject(&:double_cert)

File: lib/engine/game/g_1847_ae/share_pool.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+          shares_to_transfer = if single_shares.count >= 2
```
Comment (tobymao) [inline lib/engine/game/g_1847_ae/share_pool.rb:12]: don't use count, use size

File: lib/engine/game/g_1847_ae/step/buy_sell_par_shares.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      module Step
+        class BuySellParShares < Engine::Step::BuySellParShares
+          def can_buy?(entity, bundle)
+            return false unless super
+
+            bundle = bundle.to_bundle
+            double_cert = bundle.shares.find(&:double_cert)
+            corporation = bundle.corporation
+            # Filter out investor shares
+            ipo_shares = corporation.ipo_shares.filter(&:buyable)
```
Comment (tobymao) [inline lib/engine/game/g_1847_ae/step/buy_sell_par_shares.rb:17]: same

**Review by michaeljb — APPROVED** [review]

**Comment by Aznd4u69** [comment]
Stop this Shit!

**Pull request merged**